File: app/server/src/types/entities.py
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


def to_camel(value: str) -> str:
    parts = value.split("_")
    return parts[0] + "".join(part.capitalize() for part in parts[1:])


class CamelModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class QualityMetadata(CamelModel):
    score: float | None = None
    label: str | None = None
    source_freshness_seconds: int | None = None
    notes: list[str] = Field(default_factory=list)


class DerivedField(CamelModel):
    name: str
    value: str
    unit: str | None = None
    derived_from: str
    method: str


class HistorySummary(CamelModel):
    kind: Literal["live-polled", "propagated", "none"]
    point_count: int = 0
    window_minutes: int | None = None
    last_point_at: str | None = None
    partial: bool = False
    detail: str | None = None


class BaseEntity(BaseModel):
    model_config = ConfigDict(
        extra="allow",
        alias_generator=to_camel,
        populate_by_name=True,
    )

    id: str
    type: Literal["aircraft", "satellite", "camera", "marine-vessel", "unknown"]
    source: str
    label: str
    latitude: float
    longitude: float
    altitude: float
    heading: float | None
    speed: float | None
    timestamp: str
    observed_at: str | None = None
    fetched_at: str | None = None
    status: str
    source_detail: str | None = None
    external_url: str | None = None
    confidence: float | None = None
    history_available: bool = False
    canonical_ids: dict[str, str] = Field(default_factory=dict)
    raw_identifiers: dict[str, str] = Field(default_factory=dict)
    quality: QualityMetadata | None = None
    derived_fields: list[DerivedField] = Field(default_factory=list)
    history_summary: HistorySummary | None = None
    link_targets: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class AircraftEntity(BaseEntity):
    type: Literal["aircraft"] = "aircraft"
    callsign: str | None = None
    squawk: str | None = None
    origin_country: str | None = None
    on_ground: bool | None = None
    vertical_rate: float | None = None


class SatelliteEntity(BaseEntity):
    type: Literal["satellite"] = "satellite"
    norad_id: int | None = None
    orbit_class: str | None = None
    inclination: float | None = None
    period: float | None = None
    tle_timestamp: str | None = None


class CameraPositionMetadata(CamelModel):
    kind: Literal["exact", "approximate", "unknown"]
    confidence: float | None = None
    source: str | None = None
    notes: list[str] = Field(default_factory=list)


class CameraOrientationMetadata(CamelModel):
    kind: Literal["exact", "approximate", "ptz", "unknown"]
    degrees: float | None = None
    cardinal_direction: str | None = None
    confidence: float | None = None
    source: str | None = None
    is_ptz: bool = False
    notes: list[str] = Field(default_factory=list)


class CameraFrameMetadata(CamelModel):
    status: Literal["live", "stale", "unavailable", "viewer-page-only", "blocked"]
    refresh_interval_seconds: int | None = None
    last_frame_at: str | None = None
    age_seconds: int | None = None
    image_url: str | None = None
    thumbnail_url: str | None = None
    stream_url: str | None = None
    viewer_url: str | None = None
    width: int | None = None
    height: int | None = None


class CameraComplianceMetadata(CamelModel):
    attribution_text: str
    attribution_url: str | None = None
    terms_url: str | None = None
    license_summary: str | None = None
    requires_authentication: bool = False
    supports_embedding: bool = False
    supports_frame_storage: bool = False
    review_required: bool = False
    provenance_notes: list[str] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class CameraReviewMetadata(CamelModel):
    status: Literal["verified", "needs-review", "blocked"]
    reason: str | None = None
    required_actions: list[str] = Field(default_factory=list)
    issue_categories: list[str] = Field(default_factory=list)


class CameraEntity(BaseEntity):
    type: Literal["camera"] = "camera"
    camera_id: str | None = None
    source_camera_id: str | None = None
    owner: str | None = None
    state: str | None = None
    county: str | None = None
    region: str | None = None
    roadway: str | None = None
    direction: str | None = None
    location_description: str | None = None
    feed_type: Literal["snapshot", "stream", "page", "mixed", "unknown"] | None = None
    access_policy: str | None = None
    position: CameraPositionMetadata
    orientation: CameraOrientationMetadata
    frame: CameraFrameMetadata
    compliance: CameraComplianceMetadata
    review: CameraReviewMetadata
    health_state: str | None = None
    degraded_reason: str | None = None
    last_metadata_refresh_at: str | None = None
    next_frame_refresh_at: str | None = None
    backoff_until: str | None = None
    retry_count: int | None = None
    last_http_status: int | None = None
    nearest_reference_ref_id: str | None = None
    reference_link_status: str | None = None
    link_candidate_count: int | None = None
    reference_hint_text: str | None = None
    facility_code_hint: str | None = None


class MarineQualityMetadata(QualityMetadata):
    observed_vs_derived: Literal["observed", "derived"] = "observed"
    geometry_provenance: Literal["raw_observed", "reconstructed", "interpolated"] = "raw_observed"
    stale: bool = False
    degraded: bool = False
    source_health: str | None = None


class MarineVesselEntity(BaseEntity):
    type: Literal["marine-vessel"] = "marine-vessel"
    mmsi: str
    imo: str | None = None
    callsign: str | None = None
    vessel_name: str | None = None
    flag_state: str | None = None
    vessel_class: str | None = None
    course: float | None = None
    nav_status: str | None = None
    destination: str | None = None
    eta: str | None = None
    stale: bool = False
    degraded: bool = False
    degraded_reason: str | None = None
    source_health: str | None = None
    observed_vs_derived: Literal["observed", "derived"] = "observed"
    geometry_provenance: Literal["raw_observed", "reconstructed", "interpolated"] = "raw_observed"
    reference_ref_id: str | None = None
    quality: MarineQualityMetadata | None = None


class ReferenceObjectSummary(CamelModel):
    ref_id: str
    object_type: Literal["airport", "runway", "navaid", "fix", "region"]
    canonical_name: str
    primary_code: str | None = None
    source_dataset: str
    status: str
    country_code: str | None = None
    admin1_code: str | None = None
    centroid_lat: float | None = None
    centroid_lon: float | None = None
    bbox_min_lat: float | None = None
    bbox_min_lon: float | None = None
    bbox_max_lat: float | None = None
    bbox_max_lon: float | None = None
    coverage_tier: Literal["authoritative", "curated", "baseline"]
    object_display_label: str | None = None
    code_context: str | None = None
    aliases: list[str] = Field(default_factory=list)


class ReferenceObjectEntity(ReferenceObjectSummary):
    geometry_json: str | None = None


class AirportReferenceEntity(ReferenceObjectEntity):
    object_type: Literal["airport"] = "airport"
    icao_code: str | None = None
    iata_code: str | None = None
    local_code: str | None = None
    airport_type: str | None = None
    elevation_ft: float | None = None
    municipality: str | None = None
    iso_region: str | None = None
    scheduled_service: bool = False
    gps_code: str | None = None
    continent_code: str | None = None
    timezone_name: str | None = None
    keyword_text: str | None = None


class RunwayReferenceEntity(ReferenceObjectEntity):
    object_type: Literal["runway"] = "runway"
    airport_ref_id: str
    le_ident: str | None = None
    he_ident: str | None = None
    length_ft: float | None = None
    width_ft: float | None = None
    surface: str | None = None
    le_heading_deg: float | None = None
    he_heading_deg: float | None = None
    le_latitude_deg: float | None = None
    le_longitude_deg: float | None = None
    he_latitude_deg: float | None = None
    he_longitude_deg: float | None = None
    closed: bool = False
    lighted: bool = False
    surface_category: str | None = None
    threshold_pair_code: str | None = None
    center_latitude_deg: float | None = None
    center_longitude_deg: float | None = None


class NavaidReferenceEntity(ReferenceObjectEntity):
    object_type: Literal["navaid"] = "navaid"
    ident: str | None = None
    navaid_type: str | None = None
    frequency_khz: float | None = None
    elevation_ft: float | None = None
    associated_airport_ref_id: str | None = None
    power: str | None = None
    usage: str | None = None
    magnetic_variation_deg: float | None = None
    name_normalized: str | None = None


class FixReferenceEntity(ReferenceObjectEntity):
    object_type: Literal["fix"] = "fix"
    ident: str | None = None
    fix_type: str | None = None
    jurisdiction: str | None = None
    usage_class: str | None = None
    artcc: str | None = None
    state_code: str | None = None
    route_usage: str | None = None


class RegionReferenceEntity(ReferenceObjectEntity):
    object_type: Literal["region"] = "region"
    region_kind: Literal["country", "state", "county", "metro", "city", "custom"]
    parent_ref_id: str | None = None
    geometry_quality: str | None = None
    place_class: str | None = None
    population: int | None = None
    rank: int | None = None

File: app/client/src/types/entities.ts
export interface QualityMetadata {
  score?: number | null;
  label?: string | null;
  sourceFreshnessSeconds?: number | null;
  notes: string[];
}

export interface DerivedField {
  name: string;
  value: string;
  unit?: string | null;
  derivedFrom: string;
  method: string;
}

export interface HistorySummary {
  kind: "live-polled" | "propagated" | "none";
  pointCount: number;
  windowMinutes?: number | null;
  lastPointAt?: string | null;
  partial: boolean;
  detail?: string | null;
}

export interface TrackPoint {
  latitude: number;
  longitude: number;
  altitude: number;
  timestamp: string;
  speed?: number | null;
  heading?: number | null;
  status?: string | null;
}

export interface EntityHistoryTrack {
  entityId: string;
  entityType: "aircraft" | "satellite";
  label: string;
  semantics: "observed" | "derived";
  kind: HistorySummary["kind"];
  windowMinutes: number;
  partial: boolean;
  detail?: string | null;
  points: TrackPoint[];
}

export interface BaseEntity {
  id: string;
  type: "aircraft" | "satellite" | "camera" | "marine-vessel" | "environmental-event" | "unknown";
  source: string;
  label: string;
  latitude: number;
  longitude: number;
  altitude: number;
  heading?: number | null;
  speed?: number | null;
  timestamp: string;
  observedAt?: string | null;
  fetchedAt?: string | null;
  status: string;
  sourceDetail?: string | null;
  externalUrl?: string | null;
  confidence?: number | null;
  historyAvailable: boolean;
  canonicalIds: Record<string, string>;
  rawIdentifiers: Record<string, string>;
  quality?: QualityMetadata | null;
  derivedFields: DerivedField[];
  historySummary?: HistorySummary | null;
  linkTargets: string[];
  metadata: Record<string, unknown>;
}

export interface AircraftEntity extends BaseEntity {
  type: "aircraft";
  callsign?: string | null;
  squawk?: string | null;
  originCountry?: string | null;
  onGround?: boolean | null;
  verticalRate?: number | null;
}

export interface SatelliteEntity extends BaseEntity {
  type: "satellite";
  noradId?: number | null;
  orbitClass?: string | null;
  inclination?: number | null;
  period?: number | null;
  tleTimestamp?: string | null;
}

export interface CameraPositionMetadata {
  kind: "exact" | "approximate" | "unknown";
  confidence?: number | null;
  source?: string | null;
  notes: string[];
}

export interface CameraOrientationMetadata {
  kind: "exact" | "approximate" | "ptz" | "unknown";
  degrees?: number | null;
  cardinalDirection?: string | null;
  confidence?: number | null;
  source?: string | null;
  isPtz: boolean;
  notes: string[];
}

export interface CameraFrameMetadata {
  status: "live" | "stale" | "unavailable" | "viewer-page-only" | "blocked";
  refreshIntervalSeconds?: number | null;
  lastFrameAt?: string | null;
  ageSeconds?: number | null;
  imageUrl?: string | null;
  thumbnailUrl?: string | null;
  streamUrl?: string | null;
  viewerUrl?: string | null;
  width?: number | null;
  height?: number | null;
}

export interface CameraComplianceMetadata {
  attributionText: string;
  attributionUrl?: string | null;
  termsUrl?: string | null;
  licenseSummary?: string | null;
  requiresAuthentication: boolean;
  supportsEmbedding: boolean;
  supportsFrameStorage: boolean;
  reviewRequired: boolean;
  provenanceNotes: string[];
  notes: string[];
}

export interface CameraReviewMetadata {
  status: "verified" | "needs-review" | "blocked";
  reason?: string | null;
  requiredActions: string[];
  issueCategories: string[];
}

export interface CameraEntity extends BaseEntity {
  type: "camera";
  cameraId?: string | null;
  sourceCameraId?: string | null;
  owner?: string | null;
  state?: string | null;
  county?: string | null;
  region?: string | null;
  roadway?: string | null;
  direction?: string | null;
  locationDescription?: string | null;
  feedType?: "snapshot" | "stream" | "page" | "mixed" | "unknown" | null;
  accessPolicy?: string | null;
  position: CameraPositionMetadata;
  orientation: CameraOrientationMetadata;
  frame: CameraFrameMetadata;
  compliance: CameraComplianceMetadata;
  review: CameraReviewMetadata;
  healthState?: string | null;
  degradedReason?: string | null;
  lastMetadataRefreshAt?: string | null;
  nextFrameRefreshAt?: string | null;
  backoffUntil?: string | null;
  retryCount?: number | null;
  lastHttpStatus?: number | null;
  nearestReferenceRefId?: string | null;
  referenceLinkStatus?: string | null;
  linkCandidateCount?: number | null;
  referenceHintText?: string | null;
  facilityCodeHint?: string | null;
}

export interface MarineQualityMetadata extends QualityMetadata {
  observedVsDerived: "observed" | "derived";
  geometryProvenance: "raw_observed" | "reconstructed" | "interpolated";
  stale: boolean;
  degraded: boolean;
  sourceHealth?: string | null;
}

export interface MarineVesselEntity extends BaseEntity {
  type: "marine-vessel";
  mmsi: string;
  imo?: string | null;
  callsign?: string | null;
  vesselName?: string | null;
  flagState?: string | null;
  vesselClass?: string | null;
  course?: number | null;
  navStatus?: string | null;
  destination?: string | null;
  eta?: string | null;
  stale: boolean;
  degraded: boolean;
  degradedReason?: string | null;
  sourceHealth?: string | null;
  observedVsDerived: "observed" | "derived";
  geometryProvenance: "raw_observed" | "reconstructed" | "interpolated";
  referenceRefId?: string | null;
  quality?: MarineQualityMetadata | null;
}

export interface EarthquakeEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "usgs-earthquake";
  eventId: string;
  eventType?: string | null;
  magnitude?: number | null;
  magnitudeType?: string | null;
  depthKm?: number | null;
  place?: string | null;
  updated?: string | null;
  tsunami?: number | null;
  significance?: number | null;
  alert?: string | null;
  felt?: number | null;
  cdi?: number | null;
  mmi?: number | null;
  caveat: string;
}

export interface EonetEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "nasa-eonet";
  eventId: string;
  eventDate: string;
  categories: string[];
  statusDetail: "open" | "closed";
  geometryType: string;
  geometryCount: number;
  coordinatesSummary: string;
  description?: string | null;
  closedAt?: string | null;
  caveat: string;
}

export interface VolcanoEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "usgs-volcano-hazards";
  eventId: string;
  volcanoName: string;
  volcanoNumber: string;
  volcanoCode?: string | null;
  observatoryName: string;
  observatoryAbbr?: string | null;
  region?: string | null;
  elevationMeters?: number | null;
  alertLevel: string;
  aviationColorCode: string;
  noticeTypeCode?: string | null;
  noticeTypeLabel?: string | null;
  noticeIdentifier: string;
  statusScope: "elevated" | "monitored";
  volcanoUrl?: string | null;
  nvewsThreat?: string | null;
  caveat: string;
}

export interface TsunamiAlertEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "noaa-tsunami-alerts";
  eventId: string;
  alertType: "warning" | "watch" | "advisory" | "information" | "cancellation" | "unknown";
  sourceCenter: "NTWC" | "PTWC" | "unknown";
  issuedAt: string;
  updatedAt?: string | null;
  effectiveAt?: string | null;
  expiresAt?: string | null;
  affectedRegions: string[];
  basin?: string | null;
  region?: string | null;
  summary?: string | null;
  evidenceBasis: "advisory" | "contextual";
  caveat: string;
}

export interface UkEaFloodEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "uk-ea-flood-monitoring";
  entityKind: "alert" | "station-reading";
  eventId: string;
  severity?: "severe-warning" | "warning" | "alert" | "inactive" | "unknown" | null;
  severityLevel?: number | null;
  areaName?: string | null;
  floodAreaId?: string | null;
  riverOrSea?: string | null;
  county?: string | null;
  region?: string | null;
  message?: string | null;
  description?: string | null;
  stationId?: string | null;
  stationLabel?: string | null;
  riverName?: string | null;
  catchment?: string | null;
  parameter?: "level" | "flow" | "rainfall" | "unknown" | null;
  value?: number | null;
  unit?: string | null;
  qualifier?: string | null;
  evidenceBasis: "advisory" | "contextual" | "observed";
  sourceMode: "fixture" | "live" | "unknown";
  caveat: string;
}

export interface GeoNetHazardEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "geonet-nz";
  entityKind: "quake" | "volcano-alert";
  eventId: string;
  publicId?: string | null;
  magnitude?: number | null;
  depthKm?: number | null;
  locality?: string | null;
  region?: string | null;
  geonetQuality?: string | null;
  statusDetail?: string | null;
  volcanoId?: string | null;
  volcanoName?: string | null;
  alertLevel?: number | null;
  aviationColorCode?: string | null;
  activity?: string | null;
  hazards?: string | null;
  evidenceBasis: "observed" | "source-reported" | "advisory" | "contextual";
  sourceMode: "fixture" | "live" | "unknown";
  caveat: string;
}

export interface HkoWeatherEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "hong-kong-observatory";
  entityKind: "warning" | "tropical-cyclone-context";
  eventId: string;
  warningType?: string | null;
  warningLevel?: string | null;
  summary?: string | null;
  issuedAt?: string | null;
  updatedAt?: string | null;
  expiresAt?: string | null;
  affectedArea?: string | null;
  signal?: string | null;
  evidenceBasis: "advisory" | "contextual";
  sourceMode: "fixture" | "live" | "unknown";
  caveat: string;
}

export interface MetNoAlertEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "met-norway-metalerts";
  entityKind: "weather-alert";
  eventId: string;
  alertType: string;
  severity: "red" | "orange" | "yellow" | "green" | "unknown";
  certainty?: string | null;
  urgency?: string | null;
  areaDescription?: string | null;
  effectiveAt?: string | null;
  onsetAt?: string | null;
  expiresAt?: string | null;
  sentAt?: string | null;
  updatedAt?: string | null;
  msgType: "Alert" | "Update" | "Cancel" | "Unknown";
  statusDetail: "Actual" | "Test" | "Unknown";
  geometrySummary?: string | null;
  bboxSummary?: string | null;
  coordinateProvenance: "bbox-centroid" | "unknown";
  evidenceBasis: "advisory" | "contextual";
  sourceMode: "fixture" | "live" | "unknown";
  caveat: string;
}

export interface CanadaCapAlertEntity extends BaseEntity {
  type: "environmental-event";
  eventSource: "environment-canada-cap";
  eventId: string;
  alertType: "warning" | "watch" | "advisory" | "statement" | "unknown";
  severity: "extreme" | "severe" | "moderate" | "minor" | "unknown";
  urgency?: string | null;
  certainty?: string | null;
  areaDescription?: string | null;
  provinceOrRegion?: string | null;
  effectiveAt?: string | null;
  onsetAt?: string | null;
  expiresAt?: string | null;
  geometrySummary?: string | null;
  evidenceBasis: "advisory" | "contextual";
  sourceMode: "fixture" | "live" | "unknown";
  caveat: string;
}

export type SceneEntity =
  | AircraftEntity
  | SatelliteEntity
  | CameraEntity
  | MarineVesselEntity
  | EarthquakeEntity
  | EonetEntity
  | VolcanoEntity
  | TsunamiAlertEntity
  | UkEaFloodEntity
  | GeoNetHazardEntity
  | HkoWeatherEntity
  | MetNoAlertEntity
  | CanadaCapAlertEntity;
